### PR TITLE
gRPC: fix request context propagation

### PR DESCRIPTION
**Pull request by michalszynkiewicz**

### CODE REVIEW

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/health/GrpcHealthEndpoint.java
Comment (Ladicek) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/health/GrpcHealthEndpoint.java:17]: Should remove "and interceptor binding" now that `@GrpcEnableRequestContext` is gone?

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (Ladicek) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:34]: Are you 100% sure that the request context is never active before this line is reached? (Sorry, don't know when/from where this code is called.)

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java
Comment (Ladicek) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:51]: I think the proper sequence is `destroy` then `deactivate`. Also, why deactivate synchronously and destroy asynchronously?

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:34]: It's executed in `vertx.executeBlocking`. I think that if a request was active at the beginning of sth that is called in `executeBlocking`, it would mean the request is a leftover from some previous execution done on this worker thread.

**Comment by Ladicek** [comment]
Looks good on the first sight, but I don't truly understand how gRPC is integrated / how a gRPC request is processed to provide any meaningful review.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java
Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:51]: hmm... I took a deeper look at the code and it seems `destroy` uses the context from a thread local. And `deactivate` cleans this threadlocal... 
Given that `destroy` cleans up the context, I think it's enough to call it. 
`capturedVertxContext.runOnContext` was to make sure it is executed with the same context as the creation of the request context

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:50]: I removed `deactivate`

Comment (Ladicek) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:50]: I'd really like to have that vetted by @mkouba, because my understanding is that if you don't `deactivate`, you leave thread-local state behind.

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:50]: `deactivate` just removes the state from the current thread, it doesn't fire destroy for beans.

If the map is cleaned, why can't it be reused?

Comment (Ladicek) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:50]: Because it may be captured somewhere else for context propagation? (Not exactly sure how _that_ works.)

Comment (Ladicek) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java:50]: Ah and good point re destroying beans. In that case, IMHO both `destroy` and `deactivate` must be called.

Comment (mkouba) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java]: Both `destroy()` and `deactivate()` should be always called in the end. There's even `ManagedContext.terminate()` that should be used in simple cases. The reason why there are two separate methods is that in some situations you need to _deactivate_ the context (i.e. remove the thread local) but not _destroy_ the bean instances.

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java]: Thanks guys, I'll restore the original order to have first `deactivate` and then `destroy` then

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (cescoffier) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:39]: You would need to call event.fail on failure.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (cescoffier) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:36]: We are deactivating the context after the call, but the listener is asynchronous, so may access the request scope after the return of this method.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java
Comment (cescoffier) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:105]: Should we only activate/deactivate once (outside of the loop)?

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:36]: Got it, you wrap all the methods.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java
Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:105]: I'll try that

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java]: fixed, we have a single activate/deactivate for the whole loop now

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:42]: added

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java
Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingExecutionHandler.java:36]: this can be marked as resolved then, right @cescoffier ?

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java
Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/context/GrpcRequestContextGrpcInterceptor.java]: I think it's correct now and I'll let myself resolve this comment :)

**Review by cescoffier — APPROVED** [review]

File: integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java
Comment (Ladicek) [inline integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java:26]: This may fail intermittently, because `identityHashCode` is not guaranteed to be unique. It's not terribly likely, and I can't think of a better way, but still...

Well maybe, since this is a test, we could use an identity set (`Collections.newSetFromMap(new IdentityHashMap<>())`) and put the `ConcurrentMap` instances into it? :-) (Instead of hash codes of the `ConcurrentMap` instances.)

**Review by Ladicek — APPROVED** [review]
Looks good, baring one little comment about testing. Also, maybe squash the commits now?

(I originally asked for a separate commit to see the changes you did after our Friday call, but we should probably squash before merging.)

File: integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java
Comment (michalszynkiewicz) [reply integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java:26]: the other integration test, with normal hibernate, uses a better way of testing this, a value from a request scope bean. I thought I changed t his one too, will change it.

Comment (Ladicek) [reply integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java:26]: Ayea, that's even better!

Comment (michalszynkiewicz) [reply integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java]: done

File: integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java
Comment (Ladicek) [inline integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java:57]: Exception messages here don't correspond to present code. There's no `onCompleted` here, and there's no `RawTestService#bidi`.

Sorry to be such a pain :-)

Comment (michalszynkiewicz) [reply integration-tests/grpc-hibernate-reactive/src/main/java/com/example/reactive/ReactiveService.java:57]: Thanks for catching it, I wanted to be quick, sorry.

**Comment by gsmet** [comment]
This looks a bit too involved to get backported but I'm not sure how critical the fix is? Given there's no backport label, I understand you didn't plan to backport it?

**Comment by michalszynkiewicz** [comment]
I think it should be backported, I'll add the label.

**Review by Ladicek — APPROVED** [review]

**Review by manovotn — APPROVED** [review]
From my limited understanding PoV, this looks fairly good :-)

**Comment by gsmet** [comment]
@mkouba I let you have a look and merge it if it's fine for you.

**Comment by michalszynkiewicz** [comment]
The tests for security should not in any way be impacted by this change and they failed with downloading maven repo :| :
```
Error: The HTTP request timed out after 00:01:40.
Error: Exit code 1 returned from process: file name '/home/runner/runners/2.278.0/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.DownloadArtifact, Runner.Plugins"'.
```

**Comment by michalszynkiewicz** [comment]
@gsmet I don't think it makes sense to rerun these tests, WDYT?

**Comment by gsmet** [comment]
Agreed, merging.

**Comment by gsmet** [comment]
Ah no, I'll let @mkouba have a look if he can and merge tomorrow if we don't hear from him as it would be better to have it in 2.0.0.Final.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java
Comment (mkouba) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:68]: Hm, this may throw a `ContextNotActiveException`... Is it handled somewhere?

Comment (manovotn) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:68]: Hmm, it was my understanding that when this is called, you will always have a req. context active.

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:68]: No, the state should always be there... OTOH, if it isn't, it may not be a critical error, maybe I can just log a warning and activate a new context, WDYT?

Comment (Ladicek) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:68]: Maybe just document that assumption?

My understanding is that if `BlockingServerInterceptor` is present in the interceptor chain, it is always preceded by `GrpcRequestContextGrpcInterceptor`, and so request context should always be active.

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:68]: I'll add a comment then

Comment (michalszynkiewicz) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/blocking/BlockingServerInterceptor.java:71]: done

**Review by mkouba — APPROVED** [review]

**Pull request merged**
